sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala

```diff
@@ -156,4 +189,66 @@ class PartitionedWriteSuite extends QueryTest with SharedSparkSession {
       }
     }
   }
+
+  test("Output path should be a staging output dir, whose last level path name is jobId," +
+    " when dynamicPartitionOverwrite is enabled") {
+    withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> PartitionOverwriteMode.DYNAMIC.toString) {
+      withTable("t") {
+        withSQLConf(SQLConf.FILE_COMMIT_PROTOCOL_CLASS.key ->
+          classOf[DetectCorrectOutputPathFileCommitProtocol].getName) {
+          Seq((1, 2)).toDF("a", "b")
+            .write
+            .partitionBy("b")
+            .mode("overwrite")
+            .saveAsTable("t")
+        }
+      }
+    }
+  }
+
+  test("Concurrent write to the same table with different partitions should be possible") {
+    withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> PartitionOverwriteMode.DYNAMIC.toString) {
+      withTable("t") {
+        val sem = new Semaphore(0)
+        Seq((1, 2)).toDF("a", "b")
+          .write
+          .partitionBy("b")
+          .mode("overwrite")
+          .saveAsTable("t")
+
+        val df1 = spark.range(0, 10).map(x => (x, 1)).toDF("a", "b")
+        val df2 = spark.range(0, 10).map(x => (x, 2)).toDF("a", "b")
+        val dfs = Seq(df1, df2)
+
+        var throwable: Option[Throwable] = None
+        for (i <- 0 until 2) {
+          new Thread {
+            override def run(): Unit = {
+              try {
+                dfs(i)
+                  .write
+                  .mode("overwrite")
+                  .insertInto("t")
+              } catch {
+                case t: Throwable =>
+                  throwable = Some(t)
+              } finally {
+                sem.release()
+              }
+            }
+          }.start()
+        }
+        // make sure writing table in two threads are executed.
+        sem.acquire(2)
+        throwable.foreach { t => throw improveStackTrace(t) }
+        checkAnswer(spark.sql("select a, b from t where b = 1"), df1)
+        checkAnswer(spark.sql("select a, b from t where b = 2"), df2)
+      }
+    }
```

Comment:
Remove this UT, because dataFrame can not specify partition key value.
So with our current approach, they can not write concurrently.

Reply:
https://github.com/apache/spark/blob/a1b90bfc0faa2b4b2b7388443a734a217792d585/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L412-L420

Reply:
Or we can parse the literal partition column to static partition key value.

Reply:
P.S. unreachable code in this ut will be removed with other commits together.

Reply:
Ah, that's a limitation of `DataFrameWriter`, we may need to extend `DataFrameWriter` to support that. 

But currently, i think we can simply use the SQL syntax since we can use `spark.sql` and get the same behaviour.